src/mesh/mesh.cpp

```diff
@@ -1096,24 +1096,58 @@ void Mesh::Initialize(bool init_problem, ParameterInput *pin, ApplicationInput *
     }
 
     // send FillGhost variables
+    bool can_delete;
+    int test_iters = 0;
+    do {
+      can_delete = true;
+      for (auto &[k, comm] : boundary_comm_map) {
+        can_delete = comm.IsSafeToDelete() && can_delete;
+      }
+      test_iters++;
+    } while (!can_delete && test_iters < 1e10);
+    if (test_iters >= 1e10)
+      PARTHENON_FAIL(
+          "Too many iterations waiting to delete boundary communication buffers.");
+
     boundary_comm_map.clear();
     boundary_comm_flxcor_map.clear();
+
     for (int i = 0; i < num_partitions; i++) {
       auto &md = mesh_data.GetOrAdd("base", i);
       BuildBoundaryBuffers(md);
-      SendBoundaryBuffers(md);
     }
 
+    std::vector<bool> sent(num_partitions, false);
+    bool all_sent;
+    do {
+      all_sent = true;
+      for (int i = 0; i < num_partitions; i++) {
+        auto &md = mesh_data.GetOrAdd("base", i);
+        if (!sent[i]) {
+          if (SendBoundaryBuffers(md) != TaskStatus::complete) {
+            all_sent = false;
+          } else {
+            sent[i] = true;
+          }
+        }
+      }
+    } while (!all_sent);
```

Comment:
@jonahm-LANL's comments above regarding do-while termination here as well... can we just turn these into `for` loops with `break` statements?

Reply:
A possibly better solution to this is to create and execute a `TaskList` here, which can be generated with a single call to `AddBoundaryExchangeTasks`. I think I might save this for another PR though.

Reply:
:+1:

Reply:
Until this other PR is there, could we add counters?
I'm mostly concerned about burning computer time by sth going wrong without other processes knowing about it (and triggering a fail).

Reply:
I added some simple counters, although I am not sure what a good choice for the maximum number of times through the loop is.